skimage/transform/hough_transform.py

```diff
@@ -342,10 +351,56 @@ def hough_circle_peaks(hspaces, radii, min_xdistance=1, min_ydistance=1,
         s = np.argsort(accum / r)
     else:
         s = np.argsort(accum)
-
-    if total_num_peaks != np.inf:
-        tnp = total_num_peaks
-        return (accum[s][::-1][:tnp], cx[s][::-1][:tnp], cy[s][::-1][:tnp],
-                r[s][::-1][:tnp])
-
-    return (accum[s][::-1], cx[s][::-1], cy[s][::-1], r[s][::-1])
+    accum_sorted, cx_sorted, cy_sorted, r_sorted = \
+        accum[s][::-1], cx[s][::-1], cy[s][::-1], r[s][::-1]
+
+    tnp = len(accum_sorted) if total_num_peaks == np.inf else total_num_peaks
+
+    # Skip searching for neighboring circles
+    # if default min_xdistance and min_ydistance are used
+    if min_xdistance == 1 and min_ydistance == 1:
+        return (accum_sorted[:tnp],
+                cx_sorted[:tnp],
+                cy_sorted[:tnp],
+                r_sorted[:tnp])
+
+    # For circles with centers too close, only keep the one with
+    # the highest peak
+    removed = np.zeros(len(accum_sorted), dtype=bool)
+    accum_kept = []
+    cx_kept = []
+    cy_kept = []
+    r_kept = []
+    i = 0
+
+    # Use a KDTree to search for neighboring circles effectively
+    scaled_cx_sorted = cx_sorted / float(min_xdistance)
+    scaled_cy_sorted = cy_sorted / float(min_ydistance)
+    scaled_coordinates = np.stack([scaled_cx_sorted, scaled_cy_sorted], axis=1)
+    if scaled_coordinates.size == 0:
+        # Return empty arrays early
+        # as cKDTree throws an error given zero-sized data
+        return (np.array(accum_kept), np.array(cx_kept),
+                np.array(cy_kept), np.array(r_kept))
+    kd_tree = cKDTree(scaled_coordinates)
+    while i < tnp:
+        if not removed[i]:
+            accum_kept.append(accum_sorted[i])
+            cx_kept.append(cx_sorted[i])
+            cy_kept.append(cy_sorted[i])
+            r_kept.append(r_sorted[i])
+            # Find a short list of candidates to remove
+            # by searching within a radius of sqrt(2) in scaled space
+            neighbors_i = kd_tree.query_ball_point(
+                (scaled_cx_sorted[i], scaled_cy_sorted[i]), np.sqrt(2)
+            )
+            # Check distance in both dimensions and mark as removed if close
+            for ni in neighbors_i:
+                x_close = abs(scaled_cx_sorted[ni] - scaled_cx_sorted[i]) <= 1
+                y_close = abs(scaled_cy_sorted[ni] - scaled_cy_sorted[i]) <= 1
+                if x_close or y_close:
+                    removed[ni] = True
+        i += 1
```

Comment:
For this section, I would:

- use a for-loop instead of a while-loop. `for i in range(tnp):`
- not scale, but rather use `np.hypot(min_xdistance, min_ydistance)` for the ball tree radius, and compare the distance to `min_xdistance` and `min_ydistance` instead of 1. This removes some additional variables, and makes reasoning about the code more local.
- not create lists, but rather, use `removed` at the very end to subset the arrays with boolean masks when returning.

@xiaoyu-wu would you be happy to make these modifications? If you don't have time, I would be happy to take over for you if needed.

Reply:
I would also make this whole chunk a separate function that we call out to, to keep the hough_peaks code itself shorter and more readable.

Reply:
> use a for-loop instead of a while-loop. `for i in range(tnp):`

Here the while loop is used to stop the iteration early once the number of distant circles reach max numbers to keep. The number of iteration is unknown, which needs a while loop.

Reply:
> not scale, but rather use np.hypot(min_xdistance, min_ydistance) for the ball tree radius, and compare the distance to min_xdistance and min_ydistance instead of 1. This removes some additional variables, and makes reasoning about the code more local.

I agree that the code is easier to read this way. There is one potential pitfall when `min_xdistance` and `min_ydistance` differ largely. Each time the list returned by `query_ball_point` can include too many unnecessary candidates. But I understand it might be a very rare case.

> not create lists, but rather, use removed at the very end to subset the arrays with boolean masks when returning.

Agreed and changed as suggested.

Reply:
@xiaoyu-wu I don't understand: `i` starts at 0, is incremented by 1 each time, and is checked against a maximum value. There are no other increments of i and there is no `break` anywhere? Right now you have:

```python
i = 0
while i < max_points:
    # bunch of code that does not modify i
    i += 1
```

This is exactly equivalent to:

```python
for i in range(max_points):
    # bunch of code that does not modify i
```

... Or am I missing something?

Reply:
@jni you are right. it was a bug. It could return circles less than specified `total_num_peaks`. It wasn't caught in test because `total_num_peaks` was tested with value 1. I fixed the bug (and now using a for loop) and added another regression test.

Reply:
🙏 !

Reply:
@jni  With the latest change, I think I have addressed all your comments and the PR is ready for a final round of review. (CI needs to be restarted.)